lib/connect.js

```diff
@@ -107,12 +102,26 @@ function connect(url, socketOptions, openCallback) {
   var keepAliveDelay = sockopts.keepAliveDelay || 0;
 
   var extraClientProperties = sockopts.clientProperties || {};
-  var credentials = sockopts.credentials || credentialsFromUrl(parts);
+  
+  var protocol, fields;
+  if (typeof url === 'object') {
+	protocol = (url.protocol || 'amqp') + ':';
+	sockopts.host = url.hostname;
+	sockopts.port = url.port || ((protocol === 'amqp:') ? 5672 : 5671);
+    if (!url.username) {
+      url.username = 'guest';
```

Comment:
If `url` is mutated, perhaps it's best to clone it. This way the user won't be surprised if the value is used elsewhere.